linux/ibus-keyman/src/engine.c

```diff
@@ -150,18 +158,44 @@ ibus_keyman_engine_class_init (IBusKeymanEngineClass *klass)
     engine_class->enable = ibus_keyman_engine_enable;
     engine_class->disable = ibus_keyman_engine_disable;
 
+    engine_class->set_surrounding_text = ibus_keyman_engine_set_surrounding_text;
+    //engine_class->set_cursor_location = ibus_keyman_engine_set_cursor_location;
+
+
     engine_class->focus_in = ibus_keyman_engine_focus_in;
     engine_class->focus_out = ibus_keyman_engine_focus_out;
 
-    engine_class->page_up = ibus_keyman_engine_page_up;
-    engine_class->page_down = ibus_keyman_engine_page_down;
+    // engine_class->page_up = ibus_keyman_engine_page_up;
+    // engine_class->page_down = ibus_keyman_engine_page_down;
 
-    engine_class->cursor_up = ibus_keyman_engine_cursor_up;
-    engine_class->cursor_down = ibus_keyman_engine_cursor_down;
+    // engine_class->cursor_up = ibus_keyman_engine_cursor_up;
+    // engine_class->cursor_down = ibus_keyman_engine_cursor_down;
 
     engine_class->property_activate = ibus_keyman_engine_property_activate;
 }
 
+static void read_context(IBusEngine *engine)
+{
+    IBusKeymanEngine *keyman = (IBusKeymanEngine *) engine;
+    IBusText *text;
+    gchar *surrounding_text;
+    guint cursor_pos, anchor_pos;
+    km_kbp_context_item *context_items;
+
+    if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0)
+    {
+        g_message("reading context");
+        ibus_engine_get_surrounding_text(engine, &text, &cursor_pos, &anchor_pos);
+        surrounding_text = g_utf8_substring(ibus_text_get_text(text), 0, cursor_pos);
```

Comment:
What's the maximum length of text that can be returned here?

Reply:
Ah, the classic 'how long is a piece of string' ;)
Theoretically as long as the text widget allows it to be. `cursor_pos` is `guint` which appears to be the smallest constraint.

Reply:
OK, getting a max of 128 characters from the surrounding text